site/docs/features/gc-internals.md

```diff
@@ -0,0 +1,182 @@
+* Getting all data files for a specific content reference (think: Iceberg table snapshot)
+* Commit-log-scanning duplicate work elimination
+
+## File references
+
+All files (or objects, in case of an object store like S3) are described using a `FileReference`,
+using a _base_ URI plus a URI _relative_ to the base URI. Noteworthy: the "sweep phase", which
+"collects" all live files in a bloom filter and after that lists files in all _base_ URIs, always
+uses only the _relative_ URI, never the _full_ URI, to check whether a file is orphan or probably
+not (bloom filter is probabilistic data structure).
+
+Since object stores are the primary target, only files but not directories are supported. Object
+stores do not know about directories, further Iceberg's `FileIO` does not know about directories
+either. For file systems that do support directories this means, that empty directories will not be
+deleted, and prematurely deleting directories could break concurrent operations.
+
+## Runtime requirements
+
+Nessie GC work is dominated by network and/or disk I/O, less by CPU and heap pressure.
+
+Memory requirements (rough estimates):
+
+* Number of concurrent content-scans ("sweep phase") times the bloom-filter on-heap size
+  (assume that can be a couple MB, depending on the expected number of files and allowed
+  false-positive ratio).
+* Duplicate-commit-log-walk elimination requires some amount of memory for each distinct cut-off
+  time times the (possible) number of commits over the matching references.
+* Additional memory is required for the currently processed chunks of metadata, for example
+  Iceberg's table-metadata and manifests, to identify the live data files. (The raw metadata is
+  only read and processed, but not memoized.)
+* An in-memory live-contents-repository (**not recommended for production workloads**) requires
+  memory for all content-references.
+
+### CPU & heap pressure testing
+
+Special "tests" ([this](https://github.com/projectnessie/nessie/blob/main/gc/gc-base/src/test/java/org/projectnessie/gc/huge/TestManyObjects.java) and
+([this](https://github.com/projectnessie/nessie/blob/main/gc/gc-iceberg-inttest/src/test/java/org/projectnessie/gc/iceberg/inttest/ITHuge.java)) have
+been used to verify that even a huge amount of objects does not let a tiny Java heap "explode" and
+not use excessive CPU resources. This "test" simulates a Nessie repository with many references,
+commits, contents and files per content version. Runs of that test using a profiler proved that the
+implementation requires little memory and little CPU - runtime is largely dominated by bloom-filter
+_put_ and _maybe-contains_ operations for the per-content-expire runs. Both tests proved the
+concept.
+
+## Deferred deletion
+
+The default behavior is to immediately deletes orphan files. But it is also possible to record the
+files to be deleted and delete those later. The `nessie-gc` tool supports deferred deletion.
+
+## Non-Nessie use cases
```

Comment:
As a side note it might be worth mentioning the `Non-Nessie use cases` section on the project side... so some reshuffling of paragraphs between internal and side docs, perhaps :shrug: